### PR TITLE
[Tests-only] Add acceptance test to list all created local storage using occ command

### PR DESCRIPTION
**Pull request by jasson99**
## Description
Add acceptance test to list all created local storage using the command line.

## Related Issue

- Related #36704

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

File: tests/acceptance/features/bootstrap/OccContext.php [hunk 1]
```diff
@@ -1119,6 +1134,34 @@ public function theAdminHasAddedRemovedTheApplicableUserForMountUsingTheOccComma
 		$this->theCommandShouldHaveBeenSuccessful();
 	}
 
+	/**
+	 * @When the administrator lists all the created local storage using command line
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/OccContext.php]:
```suggestion
	 * @When the administrator lists all the created local storage using the occ command
```

File: tests/acceptance/features/bootstrap/OccContext.php [hunk 2]
```diff
@@ -1119,6 +1134,34 @@ public function theAdminHasAddedRemovedTheApplicableUserForMountUsingTheOccComma
...
+	public function theFollowingLocalStoragesShouldExist(TableNode $mountPoints)
+	{
+		$expectedLocalStorages = $mountPoints->getColumnsHash();
+		$commandOutput = json_decode($this->featureContext->getStdOutOfOccCommand());
+		foreach($commandOutput as $i) {
+			$this->createdLocalStorage[$i->mount_id] = ltrim($i->mount_point, '/');
+		}
+	foreach($expectedLocalStorages as $i) {
+	  Assert::assertContains($i['localStorage'], $this->createdLocalStorage);
+	}
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/OccContext.php]: indentation

File: tests/acceptance/features/cliLocalStorage/listLocalStorage.feature [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+Feature: create local storage from the command line
+  As an admin
+  I want to list all created local storage from the command line
+  So that I can view available local storage
+
+  Scenario: List the created local storages
+    Given the administrator has created the local storage mount "local_storage2"
+    And the administrator has uploaded file with content "this is a file in local storage" to "/local_storage2/file-in-local-storage.txt"
+    When the administrator lists all the created local storage using command line
+    Then the following local storages should exist
```
Comment (dpakach) [inline tests/acceptance/features/cliLocalStorage/listLocalStorage.feature]: Try listing local storages when there are multiple local storages mounted.

File: tests/acceptance/features/bootstrap/OccContext.php [hunk 4]
```diff
@@ -1119,6 +1134,34 @@ public function theAdminHasAddedRemovedTheApplicableUserForMountUsingTheOccComma
...
+	 *
+	 * @param TableNode $mountPoints
+	 *
+	 * @return void
+	 */
+	public function theFollowingLocalStoragesShouldExist(TableNode $mountPoints) {
+		$expectedLocalStorages = $mountPoints->getColumnsHash();
+		$commandOutput = \json_decode($this->featureContext->getStdOutOfOccCommand());
+		foreach ($commandOutput as $i) {
+			$this->createdLocalStorage[$i->mount_id] = \ltrim($i->mount_point, '/');
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/OccContext.php]: `createdLocalStorage` can be a local variable

File: tests/acceptance/features/bootstrap/OccContext.php [hunk 5]
```diff
@@ -1119,6 +1134,34 @@ public function theAdminHasAddedRemovedTheApplicableUserForMountUsingTheOccComma
...
+	 * @Then the following local storage should exist
+	 *
+	 * @param TableNode $mountPoints
+	 *
+	 * @return void
+	 */
+	public function theFollowingLocalStoragesShouldExist(TableNode $mountPoints) {
+		$expectedLocalStorages = $mountPoints->getColumnsHash();
+		$commandOutput = \json_decode($this->featureContext->getStdOutOfOccCommand());
+		foreach ($commandOutput as $i) {
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/OccContext.php]:
```suggestion
		foreach ($commandOutput as $storageEntry) {
```
Use a better variable name than `$i`

File: tests/acceptance/features/bootstrap/OccContext.php [hunk 6]
```diff
@@ -66,6 +66,11 @@ class OccContext implements Context {
 	 */
 	private $initialTechPreviewStatus;
 
+	/**
+	 * @var array $createdLocalStorage
+	 */
+	private $createdLocalStorage;
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/OccContext.php]: can be a local variable in the method

File: tests/acceptance/features/bootstrap/OccContext.php [hunk 7]
```diff
@@ -1119,6 +1134,34 @@ public function theAdminHasAddedRemovedTheApplicableUserForMountUsingTheOccComma
...
+	 *
+	 * @return void
+	 */
+	public function theFollowingLocalStoragesShouldExist(TableNode $mountPoints) {
+		$expectedLocalStorages = $mountPoints->getColumnsHash();
+		$commandOutput = \json_decode($this->featureContext->getStdOutOfOccCommand());
+		foreach ($commandOutput as $i) {
+			$this->createdLocalStorage[$i->mount_id] = \ltrim($i->mount_point, '/');
+		}
+		foreach ($expectedLocalStorages as $i) {
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/OccContext.php]:
```suggestion
		foreach ($expectedLocalStorages as $expectedLocalStorage) {
```

File: tests/acceptance/features/cliLocalStorage/createLocalStorage.feature [hunk 8]
```diff
@@ -126,4 +126,4 @@ Feature: create local storage from the command line
     # it would be helpful if they could still see "/reference1" as well as the local storage called "/reference"
     And as "user1" file "/reference1/reference-file.txt" should not exist
     #And as "user1" file "/reference1/reference-file.txt" should exist
-    #And the content of file "/reference1/reference-file.txt" for user "user1" should be "this is a reference file"
+    #And the content of file "/reference1/reference-file.txt" for user "user1" should be "this is a reference file"
```
Comment (phil-davis) [inline tests/acceptance/features/cliLocalStorage/createLocalStorage.feature]: This change can be removed

**Review by phil-davis — COMMENTED** [review]
This will be a good start.
Then we can add scenarios for local storage with different applicable users/groups etc, and check that those are listed.

File: tests/acceptance/features/bootstrap/OccContext.php (hunk 4 above)
Comment (jasson99) [reply tests/acceptance/features/bootstrap/OccContext.php]: This has been used in the other following prs as well..So I made it global..Should I make it a local variable?

Comment (phil-davis) [reply tests/acceptance/features/bootstrap/OccContext.php]: Do you need to remember it and use it in other methods later?
If so, then keep it global.
But if it is just set up (initialized...), then used in the same method, then not used again, make it local.

File: tests/acceptance/features/bootstrap/OccContext.php [hunk 9]
```diff
@@ -1119,6 +1134,34 @@ public function theAdminHasAddedRemovedTheApplicableUserForMountUsingTheOccComma
 		$this->theCommandShouldHaveBeenSuccessful();
 	}
 
+	/**
+	 * @When the administrator lists all the created local storage using occ command
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/OccContext.php]:
```suggestion
	 * @When the administrator lists the local storage using the occ command
```
This step just does the default list. So we do not need to mention "all".
(There is a command option `--all` and we will want another step one day to test that)

File: tests/acceptance/features/cliLocalStorage/createLocalStorage.feature (hunk 8 above)
Comment (phil-davis) [reply tests/acceptance/features/cliLocalStorage/createLocalStorage.feature]: Please removed this file from this PR

**Review by phil-davis — APPROVED** [review]

**Comment by phil-davis** [comment]
We can merge this when CI passes.
@jasson99 then you can rebase the other PR about deleting local storage.
And then make a PR to add scenarios that test the other "columns" of output of `files_external:list` : 
- Mount Id should have a number in it
- Storage should say "Local"
- Authentication Type should say "None"
- Configuration should have "datadir" with some string in it
- Options should be empty
- Applicable Users should be empty
- Applicable Groups should be empty

Then make some scenarios that change the "applicable users" and/or "applicable groups" and check that the list output contains those users and/or groups.

And try `files_external:list --short` - that should give less output.

**Comment by phil-davis** [comment]
Rebased just now. Maybe that will help "fix" the strange litmus CI failure.

**Pull request merged**